### PR TITLE
Add datasources

### PR DESCRIPTION
**Pull request by santoshTYSS**

### CODE REVIEW

File: Extensions/DataSources/azure-devops-extension.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  "categories": [
+    "Azure Pipelines"
+  ],
+  "targets": [
+    {
+      "id": "Microsoft.VisualStudio.Services"
+    }
+  ],
+  "icons": {
+    "default": "logo.png"
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/azure-devops-extension.json]: I retained one from samples. Shall I remove it ?

File: Extensions/DataSources/src/DataSourceHub/action-creators/DataSourceActionCreators.tsx [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    public getDataSources() {
+        var endpointId = this.getEndpointId();
+        var project = this.getProject();
+        var serviceEndpointRestClient = getClient(ServiceEndpointRestClient);
+        var data: endpointDataInfo = {}
+        serviceEndpointRestClient.getServiceEndpointDetails(project, endpointId).then((endpointdetails: ServiceEndpointDetails) => {
+            serviceEndpointRestClient.getServiceEndpointTypes(endpointdetails.type).then((typeDetails: ServiceEndpointType[]) => {
+                if (typeDetails.length == 1) {
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/action-creators/DataSourceActionCreators.tsx]: (case of git) where TYPE doesn't match and api call doesn't return anything

File: Extensions/DataSources/src/DataSourceHub/components/DataSourceResultStatus.tsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import * as React from "react";
+import * as Reflux from 'reflux';
+import { MessageCard, MessageCardSeverity } from "azure-devops-ui/MessageCard";
+
+export class DataSourceResultStatus extends Reflux.Component{
+    public render() {
+        if(this.props.statusCode=='200'){
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/components/DataSourceResultStatus.tsx]: modified

File: Extensions/DataSources/src/DataSourceHub/components/DataSourceInput.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        const DataSourceActionCreator = new DataSourceActionCreators({});
+        DataSourceActionCreator.updateDataSourceParameters(payload);
+    }
+    
+    public onSubmit(){
+        const DataSourceActionCreator = new DataSourceActionCreators({});
+        DataSourceActionCreator.executeServiceEndpointRequest(this.props.displayInfo,this.props.currentInputParam,this.props.endpointDetails);
+    }
+   
+    public createBoxes(name:string){
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/components/DataSourceInput.tsx]: modified name to renderParameters

File: Extensions/DataSources/src/DataSourceHub/components/DataSourceExecutionResult.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import * as React from "react";
+import * as Reflux from 'reflux';
+import {DataSourceError} from './DataSourceError';
+import {DataSourceResultStatus} from './DataSourceResultStatus';
+import {DataSourceResult} from './DataSourceResult'
+
+export class DataSourceExecutionResult extends Reflux.Component{
+    public render() {
+        if(this.props.result!=null){
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/components/DataSourceExecutionResult.tsx]: service endpoint request api doesnt return null which is stored in result;
here error constitutes parse error to be shown on result side

File: Extensions/DataSources/src/DataSourceHub/components/DataSourceError.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import * as React from "react";
+import * as Reflux from 'reflux';
+import { MessageCard, MessageCardSeverity } from "azure-devops-ui/MessageCard";
+
+
+export class DataSourceError extends Reflux.Component{
+    public render() {
+        if( this.props.result!=null && this.props.result.errorMessage!=''){
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/components/DataSourceError.tsx]: these errors occcurs from 
1. result object ( response from api) showing error
2. call to api itself isnt possible ( execution error)
3.parse error in datasource json

File: Extensions/DataSources/src/DataSourceHub/components/DataSourceDetails.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                <div>
+                    <div className="rhythm-vertical-16">
+                        <Header
+                        className='no-h-padding'
+                        title={"DataSource"}
+                        titleSize={TitleSize.Small}
+                        />
+                        <TextField 
+                        value= {JSON.stringify(this.props.displayInfo,null,4)}
+                        multiline
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/components/DataSourceDetails.tsx]: please elaborate

File: Extensions/DataSources/src/DataSourceHub/components/DataSourceDetails.tsx [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import { Header, TitleSize } from "azure-devops-ui/Header";
+import * as Reflux from 'reflux';
+import { TextField, TextFieldWidth } from "azure-devops-ui/TextField";
+import { DataSourceActionCreators } from "../action-creators/DataSourceActionCreators";
+
+
+export class DataSourceDetails extends Reflux.Component{
+    public onChange(e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>, item:string) {
+        if(this.props.datasourcesInfo!=null){
+            const DataSourceActionCreator = new DataSourceActionCreators({});
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/components/DataSourceDetails.tsx]: changed to let

File: Extensions/DataSources/src/DataSourceHub/action-creators/DataSourceActionCreators.tsx [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+import { endpointDataInfo, payloadInfo, paramInfo } from '../states/DataSourceExtensionState';
+
+
+export class DataSourceActionCreators extends Reflux.Component {
+
+    public updateDataSourceParameters(payload: payloadInfo) {
+        DataSourceActions.UpdateDataSourceParameters(payload)
+    }
+
+    public selectDataSource(itemtext: string|undefined, datasourceInfo: string) {
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/action-creators/DataSourceActionCreators.tsx]: removed undefined

File: Extensions/DataSources/src/DataSourceHub/DataSourceHub.json [hunk 10]
```diff
@@ -0,0 +1,21 @@
+{
+    "contributions": [
+        {
+            "id": "hub",
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/DataSourceHub.json]: chnaged to DataSources

File: Extensions/DataSources/src/DataSourceHub/DataSourceHub.html [hunk 11]
```diff
@@ -0,0 +1,7 @@
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml">
+    <body>
+        <div id="root"></div>
+        <script type="text/javascript" src="DataSourceHub.js" charset="utf-8"></script>
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/DataSourceHub.html]: changed

File: Extensions/DataSources/azure-devops-extension.json [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  },
+  "files": [
+    {
+      "path": "dist",
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/azure-devops-extension.json]: https://stackoverflow.com/questions/22842691/what-is-the-meaning-of-the-dist-directory-in-open-source-projects

File: Extensions/DataSources/azure-devops-extension.json [hunk 13]
```diff
@@ -0,0 +1,29 @@
+{
+  "manifestVersion": 1,
+  "id": "samples",
+  "publisher": "Santosh",
+  "scopes": [
+    "vso.project",
+    "vso.serviceendpoint_query"
+  ],
+  "version": "1.0.0",
+  "name": "Data Sources",
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/azure-devops-extension.json]: chnaged

File: Extensions/DataSources/azure-devops-extension.json [hunk 14]
```diff
@@ -0,0 +1,29 @@
+{
+  "manifestVersion": 1,
+  "id": "samples",
+  "publisher": "Santosh",
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/azure-devops-extension.json]: chnaged

File: Extensions/DataSources/src/DataSourceHub/action-creators/DataSourceActionCreators.tsx [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        return Parameters
+               
+    }
+    private getEndpointId(): string {
+        var currentUrl = document.referrer;
+        var regex_endpoint = /adminservices\?resourceId=(.*)/;
+        var endpointId = (currentUrl.match(regex_endpoint))![1];
+        return endpointId
+    }
```
[the remark this answers is not in the document]
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourceHub/action-creators/DataSourceActionCreators.tsx]: done

File: Extensions/DataSources/azure-devops-extension.json [hunk 16]
```diff
@@ -0,0 +1,27 @@
+{
+  "manifestVersion": 1,
+  "id": "DataSources",
+  "publisher": "ms-vscs-rm",
+  "scopes": [
+    "vso.project",
+    "vso.serviceendpoint_query",
+    "vso.gallery"
```
Comment (iashishmittal) [inline Extensions/DataSources/azure-devops-extension.json]: this is fix for the issue we are facing to access file in others machine?

File: Extensions/DataSources/azure-devops-extension.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (iashishmittal) [reply Extensions/DataSources/azure-devops-extension.json]: ok

File: Extensions/DataSources/src/DataSourcesHub/DataSourcesHub.scss [hunk 17]
```diff
@@ -0,0 +1,30 @@
+@import "node_modules/azure-devops-ui/Core/_platformCommon.scss";
+
+#datasources-root, .datasources-container{
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/DataSourcesHub.scss]: spacing before {

File: Extensions/DataSources/azure-devops-extension.json (hunk 16 above)
Comment (santoshTYSS) [reply Extensions/DataSources/azure-devops-extension.json]: Gyan is still trying with vso.gallery but didn't work

File: Extensions/DataSources/src/DataSourcesHub/Resources/DataSourcesResources.tsx [hunk 18]
```diff
@@ -0,0 +1,8 @@
+export const SUBMIT = 'SUBMIT';
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/Resources/DataSourcesResources.tsx]: make it Submit

File: Extensions/DataSources/src/DataSourcesHub/Resources/DataSourcesResources.tsx [hunk 19]
```diff
@@ -0,0 +1,8 @@
...
+export const DATA_SOURCES = 'DataSources';
+export const DATA_SOURCE = 'DataSource';
+export const SELECT_DATASOURCE = 'Select a DataSource';
+export const NO_TYPE_MATCH = 'No type match found';
+export const NO_DATASOURCE = 'No DataSources found';
+export const RESULT = 'Result';
+export const STATUS_CODE ='Status Code :';
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/Resources/DataSourcesResources.tsx]: keep Status Code here 
move logic to show : to UI

File: Extensions/DataSources/src/DataSourcesHub/Resources/DataSourcesResources.tsx (hunk 18 above)
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourcesHub/Resources/DataSourcesResources.tsx]: ok

File: Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx [hunk 20]
```diff
@@ -0,0 +1,195 @@
+import * as Reflux from 'reflux';
+import { DataSourcesActions } from '../actions/DataSourcesActions';
+import { ServiceEndpointRestClient, ServiceEndpointDetails, ServiceEndpointType, DataSource, DataSourceDetails, ResultTransformationDetails, ServiceEndpointRequest, ServiceEndpointRequestResult, OAuthConfigurationActionFilter } from 'azure-devops-extension-api/ServiceEndpoint';
+import { getClient } from 'azure-devops-extension-api/Common';
+import { DataSourceInfo,PayloadForExecuteServiceEndpointRequest,PayloadForUpdateDataSource, PayloadForUpdateDataSourceParameters, PayloadForSelectDataSource, ParamInfo } from '../states/DataSourcesExtensionState';
+
+export class DataSourcesActionCreators extends Reflux.Component{
+
+    public updateDataSourceParameters(name:string,value:string) {
+        var payload:PayloadForUpdateDataSourceParameters = {
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx]: rename PayloadForUpdateDataSourceParameters to UpdateDataSourceParametersPayload

why do we have parseError, executeError and result here?

File: Extensions/DataSources/src/DataSourcesHub/Resources/DataSourcesResources.tsx (hunk 19 above)
Comment (santoshTYSS) [reply Extensions/DataSources/src/DataSourcesHub/Resources/DataSourcesResources.tsx]: ok

File: Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+            parseError:null,
+            executeError:null,
+            result:null,
+        };       
+        DataSourcesActions.UpdateDataSourceParameters(payload)
+    }
+
+    public selectDataSource(itemtext: string, datasourceInfo: string) {
+        var Parameters:ParamInfo=this.extractParameters(datasourceInfo,null)
+        var payload:PayloadForSelectDataSource = {
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx]: rename PayloadForSelectDataSource to SelectDataSourcePayload

why do we have parseError, executeError and result here?

File: Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+            parseError: null,
+            displayInfo: datasourceInfo,
+            executeError: null
+        };
+        DataSourcesActions.SelectDataSource(payload);
+    }
+
+    public updateDataSource(newValue: string,currentInputParam:ParamInfo | null) {
+        var Parameters:ParamInfo=this.extractParameters(newValue,currentInputParam)
+        var payload: PayloadForUpdateDataSource = {
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx]: rename PayloadForUpdateDataSource to UpdateDataSourcePayload

why do we have parseError, executeError and result here?

File: Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+        };
+        DataSourcesActions.UpdateDataSource(payload);
+    }
+
+    public executeServiceEndpointRequest(DataSourceInfo: string|null, currentInputParam: ParamInfo|null, endpointDetails: ServiceEndpointDetails|null) {
+        var endpointId = this.getEndpointId();
+        var project = this.getProject();
+        var serviceEndpointRestClient = getClient(ServiceEndpointRestClient);
+        var dataSourceDetails: DataSourceDetails|null=null;
+        var resultTransformationDetails: ResultTransformationDetails | null =null;
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx]: why do we need this  | null sytnax?

File: Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+        var serviceEndpointDetails: ServiceEndpointDetails | null = null;
+        try{
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx]: format all classes 

select all text of a file and press ctrl k + ctrl f in VSCode

File: Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+            if(DataSourceInfo!=null && currentInputParam!=null && endpointDetails!=null){
+                dataSourceDetails= this.constructDataSourceDetails(DataSourceInfo,currentInputParam);
+                resultTransformationDetails= this.constructResultTransformationDetails(DataSourceInfo);  
+                serviceEndpointDetails= endpointDetails;  
+            } 
+        }
+        catch(error){
+            var payload: PayloadForExecuteServiceEndpointRequest = {
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx]: we should have seperate action for handling errors.
Store should listen to them seperately and UI should act on them seperately

File: Extensions/DataSources/src/DataSourcesHub/DataSourcesHub.tsx [hunk 26]
```diff
@@ -0,0 +1,23 @@
+import "./DataSourcesHub.scss";
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/DataSourcesHub.tsx]: In general, in overall code I see uneven usage of semicolons.
somewhere they are used and somewhere they are not used.
As in TS they are not mandatory so your code is working but we should stick to one method. (I prefer using semicolons)

Please go through all code and fix it.

File: Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+            var projectMatch = (details.split("/"));
+            if(projectMatch!=null && projectMatch.length==3){
+                      project=projectMatch[2]
+            }
+        }
+        return project;
+    }
+
+    private extractParameters(datasourceInfo: string,currentInputParam:ParamInfo | null):ParamInfo{
+        var Parameters: ParamInfo = {}
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx]: this method is little hard to understand.
lets sit together and see how can we make it easier

File: Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+            requestContent: DataSourceInfoParsed.requestContent,
+            requestVerb: DataSourceInfoParsed.requestVerb,
+            resourceUrl: DataSourceInfoParsed.resourceUrl,
+            resultSelector:DataSourceInfoParsed.resultSelector
+        };
+        return dataSourceDetails;
+    }
+
+    private constructResultTransformationDetails(dataSourceInfo:string){
+        var DataSourceInfoParsed=JSON.parse(dataSourceInfo)
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/action-creators/DataSourcesActionCreators.tsx]: dataSourceInfoParsed instaead of DataSourceInfoParsed

File: Extensions/DataSources/src/DataSourcesHub/components/DataSourceDetails.tsx [hunk 29]
```diff
@@ -1,14 +1,20 @@
 import * as React from "react";
 import { Header, TitleSize } from "azure-devops-ui/Header";
-import * as Reflux from 'reflux';
 import { TextField, TextFieldWidth } from "azure-devops-ui/TextField";
 import { DataSourceActionCreators } from "../action-creators/DataSourceActionCreators";
+import { DATA_SOURCE } from '../Resources/DataSourceResources';
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/components/DataSourceDetails.tsx]: importing individual resources is not a good practice
instead import the Resource class and then use them like Resources.DataSource

File: Extensions/DataSources/src/DataSourcesHub/Resources/DataSourceResources.tsx [hunk 30]
```diff
@@ -0,0 +1,7 @@
+export const SUBMIT = 'SUBMIT';
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/Resources/DataSourceResources.tsx]: wrap all these in Resources class
also don't name them like DATA_SOURCES, name them like DataSources

File: Extensions/DataSources/src/DataSourcesHub/components/DataSourceDetails.tsx [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+            let DataSourceActionCreator = new DataSourcesActionCreators(null);
+            DataSourceActionCreator.updateDataSource(item,this.props.currentInputParam);
+        }
+    }
+    
+    public render() {
+        if(this.props.displayInfo!=null){
+            return(
+                <div>
+                    <div className="datasource-detail-vertical-spacing-16">
```
Comment (iashishmittal) [inline Extensions/DataSources/src/DataSourcesHub/components/DataSourceDetails.tsx]: do not name your class based on the css you are applying
it should just be datasource-detail

please fix all css class names

File: Extensions/DataSources/azure-devops-extension.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (iashishmittal) [reply Extensions/DataSources/azure-devops-extension.json]: we should ask for it from Lova.

File: Extensions/DataSources/.gitignore [hunk 32]
```diff
@@ -0,0 +1,4 @@
+npm-debug.log
```
Comment (Lovakumar) [inline Extensions/DataSources/.gitignore]: We shouldn't add .gitignore .. if files are being generated here, then we might need to tweak build process a bit.

[Comment to be taken later]

File: Extensions/DataSources/package.json [hunk 33]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "datasources",
+  "version": "1.100.0",
+  "description": "Extension for Service Endpoints DataSources",
```
Comment (Lovakumar) [inline Extensions/DataSources/package.json]: Extension for Service endpoints datasources

File: Extensions/DataSources/azure-devops-extension.json [hunk 34]
```diff
@@ -0,0 +1,27 @@
+{
+  "manifestVersion": 1,
+  "id": "DataSources",
+  "publisher": "ms-vscs-rm",
```
Comment (Lovakumar) [inline Extensions/DataSources/azure-devops-extension.json]: Datasources

File: Extensions/DataSources/azure-devops-extension.json [hunk 35]
```diff
@@ -0,0 +1,27 @@
...
+  "scopes": [
+    "vso.project",
+    "vso.serviceendpoint_query",
+    "vso.gallery"
+  ],
+  "version": "1.0.553",
+  "name": "DataSources",
```
Comment (Lovakumar) [inline Extensions/DataSources/azure-devops-extension.json]: Datasources

File: azure-devops-extension.json [hunk 36]
```diff
@@ -0,0 +1,27 @@
+{
+  "manifestVersion": 1,
+  "id": "DataSources",
```
Comment (Lovakumar) [inline azure-devops-extension.json]: Datasources

File: azure-devops-extension.json [hunk 37]
```diff
@@ -0,0 +1,27 @@
...
+  "publisher": "ms-vscs-rm",
+  "scopes": [
+    "vso.project",
+    "vso.serviceendpoint_query",
+    "vso.gallery"
+  ],
+  "version": "1.0.548",
+  "name": "DataSources",
+  "description": "Extension for testing  DataSources in ServiceEndPoint",
```
Comment (Lovakumar) [inline azure-devops-extension.json]: Datasources

File: Extensions/DataSources/azure-pipelines.yml [hunk 38]
```diff
@@ -0,0 +1,16 @@
+# Azure Pipelines configuration that builds and packages the extension.
```
[the remark this answers is not in the document]
Comment (gykuma) [reply Extensions/DataSources/azure-pipelines.yml]: +1
Can we instead add a markdown file to document these build steps?

File: Extensions/DataSources/azure-pipelines-steps.yml [hunk 39]
```diff
@@ -0,0 +1,13 @@
+steps:
```
[the remark this answers is not in the document]
Comment (gykuma) [reply Extensions/DataSources/azure-pipelines-steps.yml]: +1
Can we instead add a markdown file to document these build steps?

**Pull request merged**